### PR TITLE
Preserve mtimes from archives #4392

### PR DESCRIPTION
**Pull request by arogl**
## Description

Fixes #X.  

(...)

## To Do

- [ ] Documentation. (If you've add a new command-line flag, for example, find the appropriate page under `docs/` to describe it.)
- [X] Changelog. (Add an entry to `docs/changelog.rst` near the top of the document.)
- [ ] Tests. (Encouraged but not strictly required.)

### CODE REVIEW

**Comment by arogl** [comment]
Initial change

Still to do
 [ ] test which archive type 
 [ ] how to check whether preserve_mtimes is set

**Comment by arogl** [comment]
> Is this still relevant? If so, what is blocking it? Is there anything you can do to help move it forward?
> 
> This pull request has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

Still relevant.

**Comment by arogl** [comment]
This should be an easy pull, as it mimics normal unzip behaviour of extracting files with times from the zip file.

Python unzip does not modify the files date when extracting, so it gets the current time.

File: beets/importer.py [hunk 1]
```diff
@@ -1090,6 +1090,19 @@ def extract(self):
         archive = handler_class(util.py3_path(self.toppath), mode='r')
         try:
             archive.extractall(extract_to)
+
+            # From here:
+            # https://stackoverflow.com/questions/9813243/extract-files-from-zip-file-and-retain-mod-date
+            # fixing #4392
+
+            for f in archive.infolist():
+                # path to this extracted f-item
```
Comment (sampsyo) [inline beets/importer.py:1099]: This comment doesn't seem important.

File: beets/importer.py [hunk 2]
```diff
@@ -1090,6 +1090,19 @@ def extract(self):
...
+                fullpath = os.path.join(extract_to, f.filename)
+                # still need to adjust the dt o/w item will have the current dt
+                date_time = time.mktime(f.date_time + (0, 0, -1))
+                # update date_time
+                os.utime(fullpath, (date_time, date_time))
```
Comment (sampsyo) [inline beets/importer.py:1104]: Maybe we should just put the relevant assignment closer to here?
```suggestion
                 fullpath = os.path.join(extract_to, f.filename)
                os.utime(fullpath, (date_time, date_time))
```

File: beets/importer.py [hunk 3]
```diff
@@ -1090,6 +1090,19 @@ def extract(self):
...
+
+            # From here:
+            # https://stackoverflow.com/questions/9813243/extract-files-from-zip-file-and-retain-mod-date
+            # fixing #4392
+
+            for f in archive.infolist():
+                # path to this extracted f-item
+                fullpath = os.path.join(extract_to, f.filename)
+                # still need to adjust the dt o/w item will have the current dt
+                date_time = time.mktime(f.date_time + (0, 0, -1))
```
Comment (sampsyo) [inline beets/importer.py:1102]: I don't yet understand the need for `+ (0, 0, -1)` here… in fact, I'm not entirely sure what type `f.date_time` is. Maybe this needs a closer look?

File: beets/importer.py [hunk 4]
```diff
@@ -1090,6 +1090,19 @@ def extract(self):
         archive = handler_class(util.py3_path(self.toppath), mode='r')
         try:
             archive.extractall(extract_to)
+
+            # From here:
+            # https://stackoverflow.com/questions/9813243/extract-files-from-zip-file-and-retain-mod-date
+            # fixing #4392
+
```
Comment (sampsyo) [inline beets/importer.py:1097]:
```suggestion
            # Adjust the files' mtimes to match the information from the archive. Inspired by:
            # https://stackoverflow.com/q/9813243
```

**Review by sampsyo — COMMENTED** [review]
If it works, it works!! Here are a few suggestions, especially about the comments.

File: beets/importer.py (hunk 1 above)
Comment (arogl) [reply beets/importer.py]: Comment removed

File: beets/importer.py (hunk 2 above)
Comment (arogl) [reply beets/importer.py:1107]: I have updated the code

File: beets/importer.py (hunk 3 above)
Comment (arogl) [reply beets/importer.py]: I added the question and answer from the stackoverflow link

File: beets/importer.py (hunk 4 above)
Comment (arogl) [reply beets/importer.py]: Changed as requested

**Comment by sampsyo** [comment]
Looking good! I think all this needs is to polish the comments so they're self-contained and oriented toward the code-reader of the future rather than summarizing our discussion here, and it will be good to go.

**Comment by sampsyo** [comment]
Very nice work; thanks for your effort on this!!

**Pull request merged**